Prototype: monitor disk space usage.

simonnagl opened the pull request:
This is just a prototype for disccussing some questions at this point.
 
This will fix issues #249 and #74 when implemented properly.

**Questions**
1. Should we realy implement this at proc_diskstats.c? This does not get it's values from proc. I implemented it there because the file system data is already there and it produces a graph in this section.
2. Shall we use statvfs (only mounted filesystems) or statfs (every filesystem)? If we use statfs we have to query mountinfo

**TODO**
- [x] Only add charts for filesystems where disk space is avaiable
- [x] Do not allocate and free buffer statvfs all the time
- [ ] Add this feature to the wiki
- [x] Make unit more readable (TB, GB, MB depending on filesystem size) 
- [x] Do not display disk metrics for containers, only for disks

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/firehol/netdata/338)
<!-- Reviewable:end -->

ktsaou commented:
nice!

suggestions:

In this plugin there is already a `struct disk *disk`, that includes a `struct mountinfo *`. This is used to find if the disk is a partition and get the family (mount point) name of the chart.

this is now tricky: the plugin by default ignores the partitions, but in the disk space case, we don't care if it is a partition or not. We have to assume that all cases are possible:

1. disk with mount point
2. partition with mount point
3. disk without mount point
4. partition without mount point

Also disk space monitoring should be controlled (enabled / disabled) per chart.

ktsaou commented:
updated my suggestion

simonnagl commented:
> there is already a struct disk *disk, that includes a struct mountinfo *.
Yes and no. There is a struct mountinfo used to get the family but it is not included in struct disk. But nothing we could not solve.

I think it is a bit more "tricky":
We have Disks, Containers (LVM, Crypt, ...) and Partitions.
At the moment for partitons and containers disk metrics are displayed.

We have to assume that all cases are possible, but I think we should display disk and space metrics for every type it or a object "down the tree" is mounted. For unmounted disks containers or partitions we should only display space metrics.

Why do you not want to display metrics for partitions?

ktsaou commented:
> Why do you not want to display metrics for partitions?

I just find it too much of a detail.

I have done the same in all modules. This is how the "on demand" option in enable/disable charts evolved too. Lowering the memory footprint of netdata, by default.

The user can enable them if he/she needs them.

> At the moment for partitons and containers disk metrics are displayed.

No. Everything that is **not** a partition is enabled and partitions are disabled, by default.

> We have to assume that all cases are possible, but I think we should display disk and space metrics for every type it or a object "down the tree" is mounted. For unmounted disks containers or partitions we should only display space metrics.

I don't follow you. Which tree?
I think it is simple: If a device (major:minor) is in mountinfo, it is mounted and should get a space chart. this is what `struct disk *disk` does.

Do I miss something?

simonnagl commented:
> No. Everything that is not a partition is enabled and partitions are disabled, by default.

Of course. Strange typo...

With tree I meant something similar to the output of lsblk. A partition is a child of a disk or a container. A container is child of root a container or a disk (This is different to lsblk but at lsblk containers might displayed as child of two disks. We should not do this).  A disk is the child of root.

Building such a tree would be great for more clear naming of objects. But this is another issue. I will open one when I have this one is done. I think it makes sense to disccuss it later in more detail.

Yes you are right, we should hold standard configuration simple. It should follow the following concept. *Display only charts which are usefull for every user*
Am I right?

Following this concept I think diskstats should display by default:

1. Space charts for everything mounted.
2. Metrics only for disks.

I do not know if we should display containers too by default like we do that know.
On my laptop I encrypt whoule LVM disks. For such a setup by default there is an entry for each LVM and the entry of the encrypted containter. These statistics are quite the same of course.

What do you think?

ktsaou commented:
> These statistics are quite the same of course.

Not always. I have chosen to show virtual disks (the containers to your terminology) because for many people these are some kind of RAID and in others they are special block devices, like fusion cards, etc.

Note that the virtual disks do not have all the charts of the physical disks (queued ops, merged ops, etc).

ktsaou commented:
> Following this concept I think diskstats should display by default:
> 
> Space charts for everything mounted.
> Metrics only for disks.

Yes, I agree with this.

Will you adapt this PR?

simonnagl commented:
Of course. I ping you when done.

simonnagl commented:
Done. Could you please review?

I am not quite shure where to document the changes. Could you please give me a hint?
Thank you!

ktsaou commented:
nice!

Suggestions / Questions:

1. To prevent allocating/freeing statvfs on each iteration, you could either:
 
 - add it (not as a pointer) in `struct disk`. So, each disk will have its own statvfs, or
 - define it as a static buffer at the top of the function, so it will always be there. This function is never called in parallel, so this is safe.

2. I would like to refactor it a bit to prevent the `go to` (it becomes difficult to read). Probably we could have 2 `def_enabled` variables: `def_enabled_disk` and `def_enabled_space` and then do whatever needed based on these.

3. You have commented the partition check and you have added back a case statement with hardcoded major device names. The problem with this, is that people are having proprietary disk drivers, using whatever imaginable for `major` and ask for special handling in this switch statement. This is why I added the partition check which seems to work good. Do we need to add this back?

simonnagl commented:
1. I agree. I'll use the static buffer. It reduces memory usage a bit.
2. I agree of course.
3. Mhm... The problem with using partition check is that containers will be displayed too. We discussed this above and you agreed only to display only "real" disks. lanana.org provides a standard, linux kernels should stick to. I agree that using a hard-coded list might not display every disk by default. Now I found another method to do it which is more save. We display disk without an entry in `/sys/dev/block/[maj]:[min]/slaves/`. How does this sound?

Thank you for taking your time on looking at this and making it better!

ktsaou commented:
> The problem with using partition check is that containers will be displayed too.

Can you give me an example of this?

If we use `/slaves` then all raid devices will not be displayed.

As I said, the easy decision is to enable everything by default. But the memory footprint of netdata will increase with charts that just display some detail.

I think we should disable block devices only when we are sure they provide some detail that is not that useful for most users (i.e. the users get information about the parents).

Disabling a raid device is not detail. Raids are parents of the disks. In certain setups, the disks will be the detail of the raid device.

So, I don't know. I am just trying to figure out what most users will want by default...

simonnagl commented:
Of course.

This is the output of `lsblk`
```
sda                8:0    0 465,8G  0 disk  
├─sda1             8:1    0   512M  0 part  /boot
└─sda2             8:2    0 465,3G  0 part  
  ├─STORAGE-ARCH 254:1    0    16G  0 lvm   /
  ├─STORAGE-HOME 254:2    0    16G  0 lvm   
  │ └─HOME       254:4    0    16G  0 crypt /home
  └─STORAGE-VM   254:3    0   200G  0 lvm   
    └─VM         254:5    0   200G  0 crypt /vm
```

We have one disk `sda` with two partitions. The second one `sda2` uses LVM. Two of the logical volumes are encrypted. (`HOME` and `VM`)

With the default configuration section `disks` has these entries:

- / (STORAGE-ARCH, Logical Volume)
- /home (HOME, Crypt)
- /vm (VM, Crypt)
- dm-2 (STORAGE-HOME, Logical Volume)
- dm-3 (STORAGE-VM, Logical Volume)
- sda
- sdb

The stats of `dm-2` and `/home` are almost the same. The same can be said for `dm-3` and `/vm`

I do not think that output is what the average user would expect or like.

For the setup above in my opinion the average user wants to see metrics for disks and Maby for the mounted containers. Containers between do not provide usefull detail.

> Raids are parents of the disks.

Are you shure? Are you using `mdadm` or something else. I think raid containers are always childs of partitions/disks.

> Disabling a raid device is not detail. [...]. In certain setups, the disks will be the detail of the raid device.

I do not understand what detail means here. Something the user want's to see, or something that does not give him more information about the system?

ktsaou commented:
> Are you shure? Are you using mdadm or something else. I think raid containers are always childs of partitions/disks.

Check this:

```
# lsblk 
NAME      MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda         8:0    0 931,5G  0 disk  
├─sda1      8:1    0 908,1G  0 part  
│ └─md126   9:126  0   908G  0 raid1 /
└─sda2      8:2    0  23,4G  0 part  
  └─md127   9:127  0  23,4G  0 raid1 [SWAP]
sdb         8:16   0 931,5G  0 disk  
├─sdb1      8:17   0 908,1G  0 part  
│ └─md126   9:126  0   908G  0 raid1 /
└─sdb2      8:18   0  23,4G  0 part  
  └─md127   9:127  0  23,4G  0 raid1 [SWAP]
```

or this:

```
# lsblk 
NAME      MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda         8:0    0   1,8T  0 disk  
├─sda1      8:1    0    50G  0 part  
│ └─md125   9:125  0    50G  0 raid1 /mnt/mdroot
├─sda2      8:2    0    12G  0 part  
│ └─md126   9:126  0    24G  0 raid5 
└─sda3      8:3    0   1,8T  0 part  
  └─md127   9:127  0   3,5T  0 raid5 /data
sdb         8:16   0   1,8T  0 disk  
├─sdb1      8:17   0    50G  0 part  
│ └─md125   9:125  0    50G  0 raid1 /mnt/mdroot
├─sdb2      8:18   0    12G  0 part  
│ └─md126   9:126  0    24G  0 raid5 
└─sdb3      8:19   0   1,8T  0 part  
  └─md127   9:127  0   3,5T  0 raid5 /data
sdc         8:32   0   1,8T  0 disk  
├─sdc1      8:33   0    50G  0 part  
│ └─md125   9:125  0    50G  0 raid1 /mnt/mdroot
├─sdc2      8:34   0    12G  0 part  
│ └─md126   9:126  0    24G  0 raid5 
└─sdc3      8:35   0   1,8T  0 part  
  └─md127   9:127  0   3,5T  0 raid5 /data
sdd         8:48   0 238,5G  0 disk  
├─sdd1      8:49   0     1G  0 part  /boot
├─sdd2      8:50   0    16G  0 part  [SWAP]
└─sdd3      8:51   0 221,5G  0 part  /
sde         8:64   0   1,8T  0 disk  
└─sde1      8:65   0   1,8T  0 part  
zram0     252:0    0     2G  0 disk  [SWAP]
zram1     252:1    0     2G  0 disk  [SWAP]
zram2     252:2    0     2G  0 disk  [SWAP]
zram3     252:3    0     2G  0 disk  [SWAP]
```

`md*` devices are the parent (`lsblk` shows them the other way around, but for my perception `md126` has 3 slaves, so the master info is how much disk bandwidth my `/` is using and then how much per disk).

To solve this issue and figure out what **detail** is, I suggest to think like this:

1. How much I/O my system is using? (`system.io`)
2. How much I/O my mount points are using? (charts for block devices that are mounted?)
3. How much I/O my physical disks are using? (these are the "bottlenecks", the sources of `waitio`)

Anything below the above, I believe can be considered: **detail**.

Of course I am open to any other ideas...

simonnagl commented:
Ok. I totally agree with you. I think we just had some problems understanding what the other person tries to tell. I think looking at the graph the other way round (like lsblk displays it) is easier to understand but at the end it does not make any difference.

We now want to display folowing charts by default (plugin:diskstats)

- Metrics for physical disks (devices without entries in /sys/dev/block/[maj]:[min]/slaves/)
- Metrics and space for mount points

ktsaou commented:
> We now want to display folowing charts by default (plugin:diskstats)
> 
> Metrics for physical disks (devices without entries in /sys/dev/block/[maj]:[min]/slaves/)
> Metrics and space for mount points

ok. Let's do it this way.

simonnagl commented:
Done.

It was a bit more to do.
For partitions there are no entries in slaves. I hope you like it. If you have proposals how to make the code better glad to learn and adapt it.

ktsaou commented:
ok, nice!
Give 1-2 days to review it. I am on a trip.

simonnagl commented:
No problem.

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -19,16 +22,23 @@
 	struct disk *next;
 } *disk_root = NULL;
 
+struct statvfs buff_statvfs;
```
please put this inside the function as `static`.
No major harm, just to comply with the rest of the code.

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -57,7 +67,7 @@ struct disk *get_disk(unsigned long major, unsigned long minor) {
 	d->major = major;
 	d->minor = minor;
-	d->partition_id = -1;
+	d->type = DISK_TYPE_UNKNOWN;
```
You can start with DISK_TYPE_PHYSICAL and remove the `if(d->type == DISK_TYPE_UNKNOWN)` below and the `#define DISK_TYPE_UNKNOWN`

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -69,21 +79,45 @@ struct disk *get_disk(unsigned long major, unsigned long minor) {
-		// we opened it
-		int bytes = read(fd, buffer, FILENAME_MAX);
-		close(fd);
+				d->type = DISK_TYPE_CONTAINER;
```
You can `break;` here. No need to continue the loop.
You can also write it as:

```
while((dp = readdir(dirp)))
```
and remove the `if(dp == NULL)`.

simonnagl replied:
I can not `break` here. Readdir reads a directory file by file. There is no guarantee about the order. On ext4 filesystems normaly . and .. are read first. I will now break after the first found file.

Do I realy need to add two braces around `while((dp = readdir(dirp)))`? What does that do?

ktsaou replied:
> Do I realy need to add two braces around while((dp = readdir(dirp)))? What does that do?

Without them the compiler will complain: Did the developer forgot an equal sign, or he/she really means to make dp = readdir?

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -69,21 +79,45 @@ struct disk *get_disk(unsigned long major, unsigned long minor) {
-		if(bytes > 0)
-			d->partition_id = strtoul(buffer, NULL, 10);
+	// if it is no partion neither a container it is a physical disk.
+	if(d->type == DISK_TYPE_UNKNOWN) {
```
If you remove DISK_TYPE_UNKNOWN remove this one too.

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -523,6 +678,61 @@ int do_proc_diskstats(int update_every, unsigned long long dt) {
+		if(ddo_space) {
+			if(mount_point) {
+				if (statvfs(family, &buff_statvfs) < 0) {
+					error("Faild checking disk space usage of %s", family);
```
typo

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -523,6 +678,61 @@ int do_proc_diskstats(int update_every, unsigned long long dt) {
+					space_used = (buff_statvfs.f_blocks - buff_statvfs.f_bfree) * buff_statvfs.f_bsize;
+				}
+
+				st = rrdset_find_bytype("disk_space", disk);
```
Something is wrong here.
If `statvfs()` fails, it will add the chart / refresh the values, while it should not.

Also, if all values read by `statvfs()` are zero and the `ddo_space == CONFIG_ONDEMAND_ONDEMAND` charts should not be added either.

simonnagl replied:
> If statvfs() fails, it will add the chart / refresh the values, while it should not.

Right, I will fix that.

> if all values read by statvfs() are zero

Can this realy happen? This would be a bug of statvfs? Should I add that case anyway?

ktsaou replied:
I think there are exotic filesystems out there. Even for btrfs getting the disk free space is not that simple. I guess a few filesystems will return just zero because "there cannot be a value for disk free space".

I guess this is what the ONDEMAND should do. ONDEMAND was just an idea to prevent adding charts that are always zero.

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -523,6 +678,61 @@ int do_proc_diskstats(int update_every, unsigned long long dt) {
+					} else if (space_total < (k*k*k*k*k)) {
+						unit_div = k*k*k*k;
+						unit = "TB";
+					}
```
You should always provide values with a fixed unit.

Think of it: netdata is started and you have 1TB free. It runs and collects values as TB. Then you reboot your server. netdata is restarted and now you have 500GB free. The unit is not GB, but the values already stored in netdata are TB.

netdata has an internal protection for this. If this happens, it will destroy all the previously collected data. But we should avoid it.

There is #376 that discusses how units could be changed at runtime. However, data collection should always be fixed.

simonnagl replied:
I do not calculate the unit with the amout of space free. I use total disk space.
My idea was that total disk space only changes when changing the disk. This is wrong. The size of containers and partitions can varry quite more.

Until having something that can handle to display matching units on the client side which should I use?
Megabyte or Gigabyte? I would prefer Gigabyte.

ktsaou replied:
GB is good. I agree.

ktsaou commented:
nice! the only real problem is the dynamic units adaptation you have added.
Let me know when to merge it!

Thanks!

simonnagl commented:
Thank you. I commented some of your comments again.
Thank you for helping me to improve my code!!! I hope I will not make these mistakes the next time.

If everything is ok now, please tell me. I will rebase, then you can merge.

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -69,21 +77,37 @@ struct disk *get_disk(unsigned long major, unsigned long minor) {
+
+	// find if it is a container
+	// by checking if /sys/dev/block/MAJOR:MINOR/slaves has entries
+	if(d->type != DISK_TYPE_PARTITION) {
```
minor: this is the `else` of the last `if`

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -358,6 +503,7 @@ int do_proc_diskstats(int update_every, unsigned long long dt) {
 			ddo_qops 	= config_get_boolean_ondemand(var_name, "queued operations", ddo_qops);
 			ddo_util 	= config_get_boolean_ondemand(var_name, "utilization percentage", ddo_util);
 			ddo_backlog = config_get_boolean_ondemand(var_name, "backlog", ddo_backlog);
+			ddo_space   = config_get_boolean_ondemand(var_name, "space", ddo_space);
```
I think the default value should be:

```
ddo_space   = config_get_boolean_ondemand(var_name, "space", (ddo_space && mount_point)?1:0);
```

This way, the user should know that this setting is off, by reading the config.

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -523,6 +669,40 @@ int do_proc_diskstats(int update_every, unsigned long long dt) {
+						st->isdetail = 1;
+
+						rrddim_add(st, "avail", NULL, 1, 1000*1000*1000, RRDDIM_ABSOLUTE);
+						rrddim_add(st, "reserved for root", NULL, 1, 1000*1000*1000, RRDDIM_ABSOLUTE);
```
I should have said this before:

the second argument is best not to contain spaces.
the third (shown as NULL) is the name that will appear in the dashboard.

so, you can define this as:

```
rrddim_add(st, "reserved_for_root", "reserved for root", 1, 1000*1000*1000, RRDDIM_ABSOLUTE);
```

Later, at `rrdim_set()` you should use the id: `reserved_for_root`

ktsaou commented:
Regarding documentation.

I think:

1. README.md should mention it. You can include the change in this PR.

2. If you want you can write a wiki page about "Monitoring disks with netdata". This could document the whole of this plugin.

 Usually, I prefer to use [stackedit.io](https://stackedit.io/) for writing things, which I then copy-paste to github. Keep in mind github uses its own flavor of markdown and stackedit.io has a few features (like flowcharts, UML charts, etc) which github does not have. But in general stackedit.io is super for editing markdown.

  I will also attempt to create a github repo for the wiki so that we can have PRs there too.

simonnagl commented:
> `ddo_space   = config_get_boolean_ondemand(var_name, "space", (ddo_space && mount_point)?1:0);`

I do not like this. I think this is inconsistend to how the other values are treated. On the other hand check this usecase.

Start netdata with a unmounted disk: -> `space = no`
Mount the disk -> `space = yes`
I do not think we want such a behavior.

When testing this I also noticed the behaviour is not there. We never check if a disk got mounted. I think we should check this if `space = on demand`

What do you think?

Regarding documentation:
I agree.

ktsaou commented:
> What do you think?

ok

There is an `error()` in your code, when enable is 'yes' but mountinfo is null. This will print the error every second!
If you need this error, I suggest to put a flag in `struct disk` and use it to log it just once.

simonnagl commented:
You are right. There is a similar error message with ddo_io too. I will provide try to fix this so we are able to display error messages only once here.

simonnagl commented:
Update. We need to parse /proc/mountinfo in the same interval as /proc/diskstat if we have default configuration. (and with most other configurations too)

**Reason**
If we unmount a mounted object we do not want to show statistics anymore.
Even if we would not care about that disk space would be totaly incorrect. Now the statistics of another disk would be printed.

Is there a possibility to remove graph from the dashbord if added once?

ktsaou commented:
hm... `get_disk()` caches the mountinfo linked list. I think it is overkill to read it every second.

One solution could be this:

1. `get_disk()` every time (even on cached results) updates a `time_t last_used` variable in `struct disk`.
2. `get_disk()` has a configuration option `re-read mount points when disks are lost for seconds = 5`.
3. so, in `get_disk()`, if `time(NULL) - disk->last_used > re-read mount points when disks are lost for seconds`, it free the mountinfo cache and re-read it.

How does that sound?

simonnagl commented:
Mhm...
Read it every 5 seconds will mean to approve that wrong data could be collected for 5 seconds. Parsing /proc/mountinfo will not guarantee such a failure either, but limit it to one catch. I am not happy with both solutions.

We can inhibit this failure calling `stat(mount_point)` between data collection and pushing the data to rrd. There we can check if we are still on the right disk comparing major:minor.

Checking for new mount points can then be done like you described. We then might set the default for reloading mountinfo cache higher. Maby 30 seconds?

You did not answer to my question above. Maby you missed it:
> Is there a possibility to remove graph from the dashbord if added once?
 
We should remove metrics of mounted partitions if they get unmounted. I cannot find a solution for this.

ktsaou commented:
You misunderstood it. I didn't mean to read it every 5 seconds. I meant that if **the disk is removed from the system for more than 5 seconds**, and the new disk gets the same MAJOR:MINOR, then we will discard the mountinfo cache and read it again. If it gets a new MAJOR:MINOR there will be no problem at all.

This is 100% safe. The disk is removed, so the charts have stopped collecting values. And we don't have to do anything on each iteration. Will only check if and when a new disk get's mounted with the same MAJOR:MINOR. Even the "5 seconds" should be configurable (a system may be automatically mounting/unmounting disks - although I can't think of a scenario where these disks will get the same MAJOR:MINOR).

> We should remove metrics of mounted partitions if they get unmounted. I cannot find a solution for this.

Remove means "delete"? I think we shouldn't! We are going to build alarms on this situation. Also, if we remove the metrics, the performance data of the stopped charts will be lost.

There is some code internally, that hides/disables a chart if it stops refreshing for some time. But I have disabled it. I didn't like it.

What we should do, is figure out of a way to visualize this situation. To let the user know it stopped.

simonnagl commented:
You are right. I misunderstood you. But I think you are missing one thing or I still do not get something that is  obvious for you. I know will describe the possible scenarios. Then we can discuss what to do when.

What I meant with remove was hiding the graph from the dashboard. But you are right. We should not simply remove something from the dashboard without notifying the user what happend. And why.
Until we have a feature for notifying something on the dashboard I think we should not stop collecting data of disks once added to the dashboard. In my next version here I will only stop collection of space metrics for unmounted disk. This is because I do not see a way (likely there is no way) to collect space metrics for unmounted disks.

1. A device is unmounted.
    1. If we collect space metrics we have to check if it is unmounted between data collection and pushing data to the database to ensure we do not collect wrong data. The fastest way to do this is calling `stat(mount_point)` and comparing major:minor.
    2. If we do not collect space metrics we do not need to check anything. We will continue collecting data anyway.
2. A device is mounted.
    1. We should check for new mounted devices periodically. This time can be set in conig. I found a efficient way for doing this: We check /proc/mounts for exceptions with `select()`. This indicates changes in this file. See also section `/proc/[pid]/mounts` in `man 5 proc`
    2. Only if `enable new disks detected at runtime` is set to `no` we do not do this.
3. A device is added.
    1. The right behaviour is already implemented.
4. A device is removed.
    1. We do not have to do anything. Netdata will stop monitoring this disk.

If you agree I will implement it and notify you afterwards. Thanks!

ktsaou commented:
You are right.

I didn't know `/proc/self/mounts` is pollable...

As I see it, this file is only good for polling. `/proc/self/mountinfo` has more info. So we need `/proc/self/mounts` only for detecting changes and we will still read `/proc/self/mountinfo`.

I guess this kind of polling will be needed in other places too. So I suggest to do this:

1. a new thread will take care of all filesystem pollings we need.

  we already have the main thread sitting idle (it is just there for handling signals: https://github.com/firehol/netdata/blob/0874a9852ef3e75a441e1f922111f972ebe743b6/src/main.c#L546). It could do this too.

2. for each poll it does it should have a callback function to handle the notification.

3. each callback will do whatever necessary for each plugin.

So, for the `proc_diskstats` plugin, I guess the callback will re-read mountinfo and update all the `struct disk` entries with the valid mount points.

simonnagl commented:
This sounds reasonable.

Nevertheless I would not implement such a thread in this pull request. My suggestion would be implement this for now local with `select()`. After this is merged I would implement threads for polling. I think there are some things we should discuss like if we use `poll` and one thread per polling, or if we iterate through all files and use `select`.

Will you merge this with implemented the `select()` localy and postpone a global polling unit?

ktsaou commented:
I think you cannot do it with `select()` on the `plugin_proc` thread. It will pause the thread, ie. the entire data collection.

simonnagl commented:
Shure I can. But when thinking more of it it does not make sense implementing it in proc_diskstats. There is too much overhead that we will remove soon.
Let's do it the other way round. I will open an issue to discuss global polling. When descided how to implement I can do it. Afterwards I can adapt this. Does this sound good?

ktsaou commented:
ok, let's discuss that.
I believe we have merge this in mean time though and re-factor it later.
Are there any other pending items for it?

simonnagl commented:
I added the checking with `stat` and rebased to master.
Also added to readme.

Please review and merge if it is ok for now. I'll open the discussion about the polling feature within the next days.

ktsaou commented:
nice!
Before merging it I would like to install it and test it. So give me some time.

simonnagl commented:
No problem!

simonnagl commented:
Today I implemented global polling.
You can find the current state in my fork.

How should I go on. Implement the missing features of proc_diskstats and push it together here, open a new pull request without the test changes on proc_diskstats?

I do not care. Just tell me!

ktsaou commented:
Oh! I was too busy with the registry and forgot to test/merge this.
Your call. I can try to test this later today and merge it.

simonnagl commented:
I think this would be the best solution. I will not have time for netdata today anyway.

ktsaou commented:
I checked it on my personal PC. It shows the disk space. Nice!

But, because the `/` mount point is on a partition, it enabled monitoring performance for the partition too.

I think the problem is this:

```c
		if( (d->type == DISK_TYPE_PHYSICAL) || (d->mount_point != NULL) ) {
			def_enabled = enable_new_disks;
		}
```

It enables all the charts if either the disk is physical or there is a mount for it.

I think we need 2x `def_enabled` variables. One for **disk performance** and one for **space monitoring**.

I would love to have this merged before the global polling. So I suggest this flow:

1. fix this issue
2. I will merge #405 (registry) probably later today or tomorrow
3. you will have to use the `*z()` variations of `snprintf()` and `strncpy()` - but be careful the buffer has to have an extra byte for the terminating zero.
4. we merge it
5. then, we will work on the global polling

Of course, if you prefer it in another way, let me know.

ktsaou commented:
Or, I can do the changes and merge it in #405.
Your call...

simonnagl commented:
> But, because the / mount point is on a partition, it enabled monitoring performance for the partition too.

You are right. And this is the behavior we decided above.

> - Metrics for physical disks (devices without entries in /sys/dev/block/[maj]:[min]/slaves/)
> - Metrics and space for mount points

I still think this is the behavior most people except. This answers the questions you posted:

> - How much I/O my system is using? (system.io)
> - How much I/O my mount points are using? (charts for block devices that are mounted?)
> - How much I/O my physical disks are using? (these are the "bottlenecks", the sources of waitio)

It would be nice to have some other people telling us what they would expect from a standard configuration.


The flow you proposed sounds good. I will open a pull request based on this request just now so we can start the discussion. I will simply rebase it when the two features (registry and diskstats) got merged.

ktsaou commented:
ok. I just fill it is duplicate. I have just a disk and I get 3 sets of performance counters: `sda`, `/`, `/home`. All three are about the same disk. To find out what the disk is doing I have to go to `sda` because I don't know of the `/` or the `/home` partition is taking the I/O.

Of course the user could enable any or all, but I think the default should be performance counters only for `sda`.

simonnagl commented:
I totally understand your point.
For your setup, and maybe on most one will not care about the usage performance. On the other hand I think we should show a new user that it's possible. Difficult question.

@alonbl @glensc @fredericopissarra @titpetric 
It would be cool to get your opinions too. Which disk stats we should display per default.

- Performance for physical disks and/or mounted partitions?

ktsaou commented:
In all cases I suggest to have 2 config settings, independent of each other. If you want, we can add a few more settings like:

- `enable performance metrics for physical disks = yes/no/on demand`
- `enable performance metrics for virtual disks = yes/no/on demand`
- `enable performance metrics for partitions = yes/no/on demand`
- `enable disk space monitoring = yes/no/on demand`

I just suggest the third to be `no` by default. In all cases, the settings should be separate.

simonnagl commented:
I agree with such options. Also the name `performance metrics` is quite good.
Before adding more options something else:

What meaning do you link up with `on demand`. I do not see a difference between `yes` and `on demand`. If there is no difference we should just use `yes` and `no`. This is more clear.

ktsaou commented:
> What meaning do you link up with on demand. I do not see a difference between yes and on demand. If there is no difference we should just use yes and no. This is more clear.

I think we have discussed this before:

- `yes` add the chart
- `no` don't add the chart
- `on demand` add the chart only if the values of the dimensions are not zero

`on demand` is a memory saving feature. It is all over netdata. Without it, netdata would show 2x-3x that many empty charts.

Even for disk space monitoring, where you collect 3 values (free, used, reserved), if all 3 values are just zero and the setting is `on demand`, do not show the chart.

simonnagl commented:
Yes of course. Sorry, I forgot about the `on demand` meaning. You are right, we have to use it.

Then here a base for further discussing how to do the configuration options. If we find a way to configure as much as possible in `[plugin:proc:/proc/diskstats]` do not care so much about how to set the defaults in detail.

We can separate the following (maby overlapping) disk types:
- physical disks
- virtual disks
- partitions
- mounted disks

We can display the following charts:
- Disk I/O Bandwidth (disk.*)
- Disk Completed I/O Operations (disk_ops.*)
- Disk Current I/O Operations (disk_qops.*)
- Disk Backlog (disk_backlog.*)
- Disk Utilization Time (disk_util.*)
- Average Completed I/O Operation Time (disk_await.*)
- Average Completed I/O Operation Bandwidth (disk_avgsz.*)
- Average Service Time (disk_svctm.*)
- Disk Merged Operations (disk_mops.*)
- Disk Total I/O Time (disk_iotime.*)
- Disk Space Usage (disk_space.*)

At the moment we have the following configuration options:
- enable new disks detected at runtime = on demand
- bandwidth for all disks = on demand
- operations for all disks = on demand
- merged operations for all disks = on demand
- i/o time for all disks = on demand
- queued operations for all disks = on demand
- utilization percentage for all disks = on demand
- backlog for all disks = on demand
- space usage for all disks = on demand

From the point of a user not knowing the code I would have the following questions:
- Which configuration option triggers which chart. (Maybe it would be better to have one configuration option for each chart.
- How to disable some metrics for some type of disks. (Maybe we should split the configuration of each charts in four options, one for each disk type. Only for space usage we only require one disk type (mounted)) 

I know such an approach increases the options quite much. How do you think about it?

ktsaou commented:
They are hierarchical.

1st level of options:

- `enable new disks detected at runtime`
- `enable performance metrics for physical disks`
- `enable performance metrics for virtual disks`
- `enable performance metrics for partitions`
- `enable disk space monitoring for mounted filesystems`

2nd level of options, when `enable performance metrics X` for a specific disk is `yes` or `on demand`:

- `enable I/O bandwidth`
- `enable I/O operations`
- `enable merged I/O operations`
- `enable I/O time`
- `enable queued I/O operations`
- `enable I/O utilization`
- `enable I/O backlog`

All the above calculate the default enable/disable for a chart. So, if all are `on demand`, you could disable I/O bandwidth, or I/O backlog for all disks by just swapping an option.

We could have something similar when `disk space monitoring` for a disk is `yes` or `on demand`:

- `enable disk usage`
- `enable inodes usage`
- `enable space monitoring for filesystems that are mounted multiple times` (by default `no`)

or if certain filesystems are more complex and we know how to find their space utilization:

- `enable btrfs metadata usage`
- `enable btrfs snapshots usage`

etc.

I know it is complex... we just need to find a good way of presenting it...

ktsaou commented:
I edited my response.

simonnagl commented:
Ok. Your proposal sounds good. I think I would do it a bit different but this is maybe a matter of taste. 

I think you have a fixed idea on how to do it. So I think it is quicker when you do the last commit and change the configuration options. (And include inodes statistics if you like).

Thank you for merging this.

ktsaou commented:
Don't get disappointed. I am just trying to cover as many needs as possible.
I would really love if you finish this. This is your baby.

ktsaou commented:
Regarding documentation, do you want me to add a page "Monitoring Disks" at the wiki?

simonnagl commented:
Ok. Nothing to do with disappointment.
Of course I can finish this too. A wiki entry would be great. How I told you in different posts I would like a method to contribute to the wiki for everyone very much. Writing wiki is something everybody can do and there might be some people who like to help if it is easy.

It would be also nice to have some description on the default dashboard. But I think we should do both, documentation and dashboard description after we finished the global polling and the detection of new mount points on runtime. Just because then we have to do the work once.

I will notify you when done.

ktsaou commented:
ok, nice. I'll continue discussion at #378 (documentation).

ktsaou commented:
@simonnagl I would like to make a release within the week (lots a improvements already). I would love to have this included. Do you believe you can do it?

simonnagl commented:
No. I do not think we get this merged in the next week. At the moment I try to write some documentation to understand what the graphs display and how this can be used. I do not like having this in the next release in this state. Maybe we should postpone the enhancements of diskstats to the next release for this reasons.

1. No pressure on developement
2. If we merge this we will change diskstats in the next release again (polling, bugfix of #393)

simonnagl commented:
I did some research on disk performance. My intention was to write some text to describe how the charts should be used to gather information about the disk. I will split my post into two section.

1. I will describe the current charts, what they display and how they can be used.
2. I propose a different order and change some charts.

Current diskstat charts
================

### Disk I/O Bandwidth (`disk.*`)

Displays the kilobytes per written/read to/from the disk.

You can use this graph to see how much data was written and read from your disk.

### Disk Completed I/O Operations (`disk_ops.*`)

Number of read and write operations per second actually done on disk.

You can see the operations done on disk. Multiplied with the value from merged operations you will get the number of operations requested from software.

### Disk Current I/O Operations (`disk_qops.*`) Number of I/O operations

Operations currently in progress. This is always a snapshot, not an average value. This indicates how many operations are currently in progress but not finished.

A high value indicates that the disk has many open operations at the time reading the value. 

Maybe we should remove this because it is only a snapshot. The value under `disk_balcklog.*` represents the same thing, but without the inaccuracy due to being a snapshot.

### Disk Backlog (`disk_backlog.*`) 

weighted # of milliseconds spent doing I/Os This field is incremented at each
I/O start, I/O completion, I/O merge, or read of these stats by the number of
I/Os in progress (field 9) times the number of milliseconds spent doing I/O 
since the last update of this field.  This can provide an easy measure of both
I/O completion time and the backlog that may be accumulating.

### Disk Utilization Time (`disk_util.*`) number of milliseconds spent doing

number of milliseconds spent doing I/Os. This field increases so long as field 9
is nonzero.

### Average Completed I/O Operation Time (`disk_await.*`) 

The averange of time needed for a read/write operation.

### Average Completed I/O Operation Bandwidth (`disk_avgsz.*`) 

The averange of kilobytes written or read per operation.

### Average Service Time (`disk_svctm.*`)

The averange of buisy time needed for a read/write operation.

### Disk Merged Operations (`disk_mops.*`)

Number of reads or writes merged. Reads and writes which are adjacent to each
other may be merged for efficiency.  Thus two 4K reads may become one 8K read
before it is ultimately handed to the disk, and so it will be counted (and
queued) as only one I/O.  This field lets you know how often this was done.

### Disk Total I/O Time (`disk_iotime.*`)

This is the total number of milliseconds spent by all reads or writes.

This value is around 40.000 on my laptop if I do many read or write operations. This also means the sum of amounts of `disk_avgsz.*` is much bigger then 10 times disk_svctm.
@ktsaou Do you know how this makes sense? How to read this values?

Proposed diskstat charts
================

### Disk I/O Bandwidth (`disk.*`)

### Disk Completed I/O Operations (`disk_ops.*`)

### Disk Merged Operations (`disk_mops.*`)

### Disk Backlog (`disk_backlog.*`) 

@ktsaou I do not like the name, because it does not display how long it will take to finish running I/O operations. Sadly I do not know a better one.

### Disk Utilization Time (`disk_util.*`) number of milliseconds spent doing

### Average Service Time (`disk_svctm.*`)

### Average Completed I/O Operation Bandwidth (`disk_avgsz.*`) 

### Disk Total I/O Time (`disk_iotime.*`)

### Average Completed I/O Operation Time (`disk_await.*`) 

### Disk Space Usage

### Disk Inodes Usage

ktsaou commented:
When I wrote these charts I first tried to understand what `vmstat` and `iostat` are presenting. I walked through their manual pages, their source code and searched the net for questions about each metric.

It is important to realize that there is a plethora of different disk types out there (magnetic, ssd, raids, fusion cards, memory cards, ram drives, even optical and tape drives)  and each chart is supposed to help understand what is really happening in all cases.

I suggest that we should just describe what each chart measures, as precisely as possible, without taking any conclusions on what is good or bad. Take for example issue #288, this fellow tried to understand what 100% disk utilization means for him. He has a fusion card. For a fusion card, the utilization chart actually means nothing. Should we remove the chart? No, because it is useful in many other cases.

I also suggest that we describe all charts at `index.html` (there is an array there just for that purpose - check #279 ), so that all the information about the charts is right on the dashboard with the charts (If you want help achieving this, I can show you how). Documentation on the wiki/docs should only exist for "configuring disk monitoring".

ktsaou commented:
Of course, we could also right an article "Monitoring disks with netdata" which could describe the whole idea, the different charts, etc.

simonnagl commented:
Yes, you are right. Removing that chart is not what we want then.
For the next step I will try to document the chats in index.html. I will ask you for help if I have problems.

simonnagl commented:
I added

- Descriptions to the dashboard. Pleas review and give hints if you think we could do int better.
- Enhance options
- Added Inodes chart. TODO:
    1. We should handel `on demand` the right way.
    2. Should we display the value in thousands? What do you think?

ktsaou commented:
I am testing it. Very nice work!

Observations:

1. There are 2 compilation warnings (`-Wall -Wextra`).
2. The info texts you added at index.html I think they need some work. I'll make comments on the code of your PR.
3. Regarding the settings, I would also enable virtual disks by default (virtual disks will only render bandwidth and IOPs due to `on demand`, since all their other metrics are zero - it is good to have bandwidth and IOPs for your RAID5).

Veeeeeeeeeeery nice work!
Thank you for taking the time to do this!

ktsaou commented on web/index.html:
```diff
@@ -1192,7 +1192,7 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 	'disk': {
 		title: 'Disks',
-		info: undefined
+		info: 'All metrics, except space and inodes, are read from <code>/proc/diskstats</code>. The field numbers mentioned in the subsection refer to the field numbers in the <a href="https://www.kernel.org/doc/Documentation/iostats.txt">Kernel Documentation</a>. Space metrics are read via system call <code>statvfs()</code>'
```
I suggest the following text:

Charts with performance information for all the system disks. Special care has been given to present disk performance metrics in a way compatible with <code>iostat -x</code>. netdata by default prevents rendering performance charts for individual partitions. Partitions can still be enabled by altering the relative settings in the netdata configuration file.

simonnagl replied:
We should do this when the page is there. I will include the proposed text.

ktsaou replied:
I added some stub text on that page.
Do you want to complete it or you want me to do it?

simonnagl replied:
Please do so if you like. I would like to focus on the polling thing if this is mergeable now.

ktsaou commented on web/index.html:
```diff
@@ -1435,18 +1435,29 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 			gaugeChart('Utilization', '12%', '', '#FF5588')
-		]
+		],
+		info: 'Percent of time the disk was doing something. Read from field 10.'
```
I suggest this text:

Disk Utilization measures the amount of time the disk was busy with something. This is not related to its performance. 100% means that the Linux kernel always had an outstanding operation on the disk. Keep in mind that depending on the underlying technology of the disk, 100% here may or may not be an indication of congestion.

ktsaou commented on web/index.html:
```diff
@@ -1435,18 +1435,29 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 	'disk.backlog': {
-		colors: '#0099CC'
+		colors: '#0099CC',
+		info: 'Milliseconds spent doing I/Os multiplied by the number of operations currently in progress. Read from field 11.'
```
I suggest the same text from kernel documentation:

Backlog is an indication of the duration of pending disk operations. So, while the disk is busy the Linux kernel is multiplying the time spent doing I/O with the number of pending operations. While not accurate, this metric can provide an indication of the expected completion time of the operations in progress.

simonnagl replied:
Here I kept the text short because:
- The value is no accurate backlog.
- For disks which are able to do operations in parallel this value is a multiple of the "real" backlog.

I suggest this text:

Backlog is an indication of the duration of pending disk operations. On every I/O event the Linux kernel is multiplying the time spent doing I/O since the last update of this field with the number of pending operations. While not accurate, this metric can provide an indication of the expected completion time of the operations in progress.

ktsaou commented on web/index.html:
```diff
@@ -1435,18 +1435,29 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 			gaugeChart('Write', '12%', 'writes')
-		]
+		],
+		info: 'Amount of data transfered to and from disk. Computed from fields 3 and 7.'
```
Amount of data transferred to and from disk.

(fixed a typo)

ktsaou commented on web/index.html:
```diff
@@ -1435,18 +1435,29 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
+	},
+
+	'disk.ops': {
+		info: 'Number of operations done on disk. The operations requested to the disk might be higher because of merged operations. Read from fields 1 and 5.'
```
Completed disk I/O operations. Keep in mind the number of operations requested might be higher, since the Linux kernel is able merge adjacent to each other (see merged operations chart).

ktsaou replied:
I updated it

ktsaou commented on web/index.html:
```diff
@@ -1435,18 +1435,29 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
+	},
+
+	'disk.qops': {
+		info: 'Number of operations currently in progress. This value is a snapshot, no average over the last interval. Read from field 9.'
```
I/O operations currently in progress. This metric is a snapshot - it is not an average over the last interval.

ktsaou commented on web/index.html:
```diff
@@ -1463,19 +1474,24 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 	'disk.iotime': {
-		height: 0.5
+		height: 0.5,
+		info: 'Number of milliseconds spend by all operation. Read from fields 4 and 8.'
```
The sum of the duration (in milliseconds) of all completed I/O operations. This number can exceed the interval if the disk is able to execute I/O operations in parallel.

ktsaou commented on web/index.html:
```diff
@@ -1463,19 +1474,24 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 	'disk.mops': {
-		height: 0.5
+		height: 0.5,
+		info: 'Number of merged operations. Read from fields 2 and 6.'
```
The number of merged disk operations. The Linux kernel is able to merge adjacent I/O operations, for example two 4KB reads can become one 8KB read before it is handled to the disk.

ktsaou commented on web/index.html:
```diff
@@ -1463,19 +1474,24 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 	'disk.svctm': {
-		height: 0.5
+		height: 0.5,
+		info: 'Buisy time devided through the number of operations.'
```
The average service time (in milliseconds) for completed I/O operations. This metric is calculated using the total busy time of the disk and number of completed operations. If the disk is able to execute multiple parallel operations the reporting average service time will be misleading.

ktsaou replied:
updated it

ktsaou commented on web/index.html:
```diff
@@ -1463,19 +1474,24 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 	'disk.await': {
-		height: 0.5
+		height: 0.5,
+		info: 'Averange of time needed for a operation.'
```
The average time (in milliseconds) for I/O requests issued to the device to be served. This includes the time spent by the requests in queue and the time spent servicing them.

(taken from the iostat manpage)

ktsaou commented on web/index.html:
```diff
@@ -1463,19 +1474,24 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 	'disk.avgsz': {
-		height: 0.5
+		height: 0.5,
+		info: 'Averange of data written or read per operation.'
```
The average size, in kilobytes, per I/O operation.

simonnagl commented:
Another round. Thanks for you patience.

> There are 2 compilation warnings (-Wall -Wextra).

Done

> The info texts you added at index.html I think they need some work. I'll make comments on the code of your PR.

Thanks for you help. I removed units form the description. They may change in later versions (#376) and are redundant on the dashboard.

> Regarding the settings, I would also enable virtual disks by default (virtual disks will only render bandwidth and IOPs due to on demand, since all their other metrics are zero - it is good to have bandwidth and IOPs for your RAID5).

This is not correct for all virtual disks. For `LVM` logical volumes there are all metrics available.
Also, on my laptop enabling virtual disks would enable many redundant charts. I use `crypt` over `lvm` to encrypt some logical volumes. Enabling virtual disks enables
1. The charts for the crypt virtual disk.
2. The charts for the underlying logical volume.
3. The charts for the `lvm` volume group.

ktsaou commented on web/index.html:
```diff
@@ -1435,18 +1435,29 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
+	},
+
+	'disk.ops': {
+		info: 'Completed disk I/O operations. Keep in mind the number of operations requested might be higher, since the Linux kernel is able merge adjacent to each other (see merged operations chart).'
```
... is able **to** merge **operations** adjacent to each other ...

(my bad)

ktsaou commented on src/proc_diskstats.c:
```diff
@@ -69,21 +77,35 @@ struct disk *get_disk(unsigned long major, unsigned long minor) {
+				}
+				d->type = DISK_TYPE_CONTAINER;
+				// Stop the loop after we found one file.
+				break;
```
instead of breaking the loop, we could count the number of slaves and instead of `d->type` we could have `d->slaves`.

why?

We could use the number of slaves to decide if we should enable by default the virtual disk or not. Virtual disks with just one slave can be disabled by default, while virtual disks with more than one slave can be enabled by default.

Will this solve your virtual disks case while allowing RAID devices to be enabled by default?

simonnagl replied:
> instead of d->type we could have d->slaves.

This won't work. Then we cannot distinguish between physical and virtual disks. This might not be relevant for default configuration but for the configuration options we like to provide.

Anyways counting `d->slaves` would solve my `LVM` setup, but I do not know if it is worthwhile to do it.
Imagine a `lvm` setup on one disk with various logical partitions.

There the logical volume group and the disk will be monitored and displayed. Because the group does only hold one disk the charts are quite the same.

To do the feature you suggest the right way we would need to find out all 1-1 relationships between "disks" and only display one of it. 

I can implement this if you want.
Personally I would not do such much effort on the default configurations. There are too many things we might miss and with the configurations options we have now most things can be configured with a few lines.

What do you think?

simonnagl replied:
Edited the comment above.

ktsaou replied:
I have never used LVM, so please excuse my ignorance.

The only problem I see is RAID devices, for which we will not show their performance charts by default. People will think netdata is not able to monitor them.

Hm... another solution could be to show virtual devices that have slaves, only if they are also mounted. We will lose swap devices on RAID, but I think we can live with that.

Would this solve the problem?

simonnagl replied:
Yes, this sounds great. I will provide a fix later on.
The question is how we represent this on the configuration.
I think we need another configuration option `space metrics for mounted virtual disks`.

ktsaou commented:
Very nice. It looks pro!

ktsaou commented on web/index.html:
```diff
@@ -1192,7 +1192,7 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 	'disk': {
 		title: 'Disks',
-		info: undefined
+		info: 'Charts with performance information for all the system disks. Special care has been given to present disk performance metrics in a way compatible with <code>iostat -x.</code> netdata by default prevents rendering performance charts for individual partitions. Partitions can still be enabled by altering the relative settings in the netdata configuration file.'
```
... <code>iostat -x</code>. netdata by default prevents rendering performance charts for individual partitions **and virtual disks with just one physical disk**.

(subject to the change I proposed above)

ktsaou replied:
(also move the dot outside the `code` block - it seems like it is part of the `iostat -x` command.

ktsaou commented:
We could also add info for space and inodes:

**space**

Disk space utilization. `reserved for root` is automatically reserved by the system to prevent the `root` user from getting out of space.

**inodes**

inodes (or index nodes) are filesystem objects (e.g. files and directories). On many types of file system implementations, the maximum number of inodes is fixed at filesystem creation, limiting the maximum number of files the filesystem can hold. It is possible for a device to run out of inodes. When this happens, new files cannot be created on the device, even though there may be free space available.

(inodes description taken from wikipedia: https://en.wikipedia.org/wiki/Inode#Details)

simonnagl commented:
Done. Please also see my comment above.
Anything else?

simonnagl commented:
Done. Changed the default configuration again.
Anything else?

ktsaou commented on web/index.html:
```diff
@@ -1463,21 +1474,33 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
 
+	'disk.space': {
+		info: 'Disk space utilization. reserved for root is automatically reserved by the system to prevent the root user from getting out of space.'
+	}
```
there a comma missing here

ktsaou commented:
shall we merge it? (after adding the missing comma).

ktsaou commented on web/index.html:
```diff
@@ -1463,21 +1474,33 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
+		info: 'Disk space utilization. reserved for root is automatically reserved by the system to prevent the root user from getting out of space.'
+	}
+	'disk.inodes': {
+		'inodes (or index nodes) are filesystem objects (e.g. files and directories). On many types of file system implementations, the maximum number of inodes is fixed at filesystem creation, limiting the maximum number of files the filesystem can hold. It is possible for a device to run out of inodes. When this happens, new files cannot be created on the device, even though there may be free space a ...
```
and `info: ` missing here

ktsaou commented on web/index.html:
```diff
@@ -1463,21 +1474,33 @@ <h4 class="modal-title" id="switchRegistryModalLabel">Switch Netdata Registry Id
+	}
+	'disk.inodes': {
+		'inodes (or index nodes) are filesystem objects (e.g. files and directories). On many types of file system implementations, the maximum number of inodes is fixed at filesystem creation, limiting the maximum number of files the filesystem can hold. It is possible for a device to run out of inodes. When this happens, new files cannot be created on the device, even though there may be free space a ...
+	}
```
and another comma here

ktsaou commented:
I tested with valgrind. It is good.
You have done a perfect work. The result is very professional. Thank you!

Let me know when to merge it.

simonnagl commented:
I think too you can merge it. Thank you very very much for your help. It does make much fun to work on that project.
After you merged it I will rebase the polling pull request and notify you when I think it is done.

ktsaou commented:
merged it.

ktsaou commented:
I fixed the broken index.html (I thought you had fixed it - by bad). Anyway it is ok now.

ktsaou commented:
On one server I see the following issue:

I have an ssd disk `/dev/sdd`. `/dev/sdd3` is mounted as `/`, but disk space monitoring is not enabled by default.

Here is the relative config (by default):

```
[plugin:proc:/proc/diskstats:sdd3]
	# enable = on demand
	# enable performance metrics = no
	# enable space metrics = no
```

ktsaou commented:
I looking at it. Don't worry...

ktsaou commented:
I will make a new PR. There are too many `config_get()` call executed for each disk on each iteration. I'll attempt to cache them in `struct disk` and rename the various variables so that we will be able to work further on this module easily.

uniquegodwin commented:
Does it show the disk usage ie space used/left? Im looking for that feature. It would be very helpful if thats available. Thanks.

ktsaou commented:
> Does it show the disk usage ie space used/left? Im looking for that feature. It would be very helpful if thats available. Thanks.

of course, disk space and inodes: http://london.my-netdata.io/#menu_disk_submenu__
there are also many alarms related to this.
